core/src/multihash.rs

```diff
@@ -0,0 +1,34 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use multihash::{self, Multihash as MultihashGeneric, U32, U64, typenum::U42};
+use multihash::derive::Multihash;
+pub use multihash::MultihashDigest;
+
+pub type Multihash = MultihashGeneric<U64>;
+
+#[derive(Copy, Clone, Debug, Eq, Multihash, PartialEq)]
+#[mh(alloc_size = U64)]
+pub enum Code {
+    #[mh(code = 0x00, hasher = multihash::IdentityHasher::<U64>, digest = multihash::IdentityDigest<U42>)]
+    Identity,
+    #[mh(code = 0x12, hasher = multihash::Sha2_256, digest = multihash::Sha2Digest<U32>)]
+    Sha2_256,
+}
```

Comment:
I don't understand why `multihash` has been changed to require this type of glue code?

All we want is a small crate that provides a glorified `struct Multihash(Code, Vec<u8>); enum Code { Identity, Sha2 }`, not (to exaggerate) an entreprise framework.

Reply:
If you only want a glorified `struct Multihash(Code, Vec<u8>)`, you don't need to do that, you could just import `Multihash` instead. Though then you would lose the ability to do the hashing of data, it would need to be done separately. I can also do a PR to see what that would look like.

The reason to do it this way (with a custom table) is that with the default Multihash table, you would also pull in hash algorithm dependencies you never need. Also the default table doesn't support the Identity hash (long story short, Identity hash is too often not a good idea, hence it's not part of it by default (in libp2p it's OK though)).

Reply:
I just had an idea that might work for libp2p to make this way simpler, let me have a try an I come back to you to see if that fits your needs better.

Reply:
> I don't really understand why `multihash` has been changed to require this type of glue code?

`multihash` is now a bit more low level and more like a building block for your application. If you don't want to pull in several hashers you don't even need, you need to create your own table.

I've pushed a commit which makes things more explicit and is not using the macro. I hope that glue is small enough and also makes it clearer why it is needed.

I tried a simpler version, with just having constants for `IDENTITY` and `SHA2_256` and a `digest` method, but that didn't really end up nicely. The current code is way more type safe (though again more similar to the original code).

Reply:
And yet another idea. In `multihash` the default table has a feature flag for each of is codec. I thought it wasn't useful as you would either want to experiment with `multihash`, hence just use the default table and don't care about the additional dependencies. I further thought that if you only need a subset, you'd use the derive macro or implement it yourself.

Looks like I was wrong.

Let me have yet another look to see if I can revive those feature flags in a good way and whether libp2p could leverage that feature.

Reply:
Don't take this wrong, but at some point I don't really see the point of having a dependency anymore. I have the feeling that the code would be more simple if we did the prefixing and hashing manually.

Reply:
Thanks for the feedback, that's exactly what I was looking for. I don't want to create libraries that don't cover real world use cases. So let's see what I can come up with in `multihash` to make it really usable for libp2p.

Reply:
@tomaka I pushed a new branch of rust-multihash. Does this look better now (I like it a lot better)?

Reply:
:+1: